### ISSUE TITLE
ERROR: Error extracting LR logs: 1 validation error for DynamicSchemaModel

### ISSUE THREAD

**Issue by brunomgalmeida** [dgea005/pypgoutput#13]
Hi,

I am playing around with your code and I came accross this error. 

```py
ERROR: Error extracting LR logs: 1 validation error for DynamicSchemaModel
```

You can reproduce with 

```sql
DROP TABLE public.bets1;
DROP PUBLICATION test_pub;
select pg_drop_replication_slot('test_slot');

CREATE PUBLICATION test_pub FOR ALL TABLES;
SELECT * FROM pg_create_logical_replication_slot('test_slot', 'pgoutput');

CREATE TABLE if not exists public.bets1 (
    id bigint NOT NULL,
    deleted numeric(1,0) DEFAULT (0)::numeric NOT NULL,
    oca smallint DEFAULT 0 NOT NULL,
    modified timestamp with time zone DEFAULT now(),
    reference character varying(250),
    status character varying(20) DEFAULT 'ACTIVE'::character varying NOT NULL,
    bet_type character varying(20) NOT NULL,
    placed_time timestamp with time zone DEFAULT now() NOT NULL,
    total_stake numeric(10,2) NOT NULL,
    account_ref character varying(50) NOT NULL,
    entered_by_user_id bigint,
    settled_by_user_id bigint,
    settlement_time timestamp with time zone,
    refund numeric(10,2),
    winnings numeric(10,2),
    channel character varying(20) NOT NULL,
    currency character varying(3) NOT NULL,
    bets_id bigint,
    chan_id bigint NOT NULL,
    acco_id bigint NOT NULL,
    risk_hit numeric(1,0) DEFAULT (0)::numeric NOT NULL,
    bet_risk_group character varying(50),
    rejection_details character varying(100),
    rejection_checked numeric(1,0) DEFAULT (0)::bigint NOT NULL,
    rejected numeric(1,0) DEFAULT (0)::numeric NOT NULL,
    total_allowed_stake numeric(10,2),
    void_reason character varying(100),
    system_bet numeric(1,0) DEFAULT 0 NOT NULL,
    bet_type_id smallint,
    status_id smallint NOT NULL,
    total_price numeric(26,4) NOT NULL,
    message_time timestamp with time zone,
    external_bet_id character varying(30),
    settlement_type smallint,
    win_amt_adjust numeric,
    place_amt_adjust numeric,
    num_bets smallint NOT NULL,
    notes character varying(200),
    manual_payout numeric,
    partition_no numeric(1,0) NOT NULL,
    payout numeric(10,2),
    prod_id bigint,
    setl_type_id smallint,
    captured_stake numeric(10,2),
    win_type_id smallint,
    agent_positions character varying(100),
    agent_commissions character varying(100),
    account_path character varying(80),
    account_bonuses_id bigint,
    free_bet boolean,
    bonus_campaign_id bigint,
    country_code_iso character(3),
    bonus_stake numeric(10,2),
    operator_stake numeric(10,2),
    operator_liab numeric(10,2),
    betslip_id bigint,
    wallet_error_code character varying(30),
    wallet_error_details character varying(200),
    wallet_type_id smallint NOT NULL,
    fund_type_id smallint NOT NULL,
    betslip_settled boolean,
    num_lines smallint NOT NULL,
    void_by_user_id bigint,
    punter_colour character varying(20),
    trading_period_id smallint,
    regulator_conf_id character varying(20),
    site character varying(5) NOT NULL,
    enhancement1_type character varying(10),
    enhancement1_flag boolean,
    enhancement1_fee numeric(10,2),
    enhancement1_multiplier numeric(10,2),
    tax_on_returns_pct numeric,
    tax_on_returns_amt numeric,
    winnings_bonus numeric(10,2),
    betslip_settled_by bigint,
    tax_on_stake_pct numeric,
    tax_on_stake_amt numeric,
    bonus_deposit_stake numeric(10,2),
    bonus_winning_stake numeric(10,2),
    bonus_funds_stake numeric(10,2),
    over_ask_id bigint,
    over_ask_expiry timestamp with time zone,
    limited_account boolean,
    affiliate_id character varying(255),
    result smallint DEFAULT 0 NOT NULL,
    pct_max_stake_used numeric(7,4),
    wallet_settlement_status_id smallint,
    comment_time timestamp with time zone,
    commented_by_user_id smallint,
    click_thru_source character varying(20),
    payout_excluding_bog numeric(10,2),
    entitled_free_bet boolean,
    external_betslip_id character varying(30),
    org_cashout_bet_id bigint,
    cashout_count smallint,
    partial_cashout numeric(1,0),
    cashout_amt numeric(10,2),
    replacement_bet_id character varying(50),
    build_a_bet boolean,
    blocked boolean DEFAULT false,
    max_payout numeric(10,2),
    autocashout character varying(2000),
    referral_path character varying(25),
    external_retailer_id bigint,
    retail_bet_info character varying(2000),
    claim_by_date timestamp with time zone,
    paid_time timestamp with time zone,
    paid_by_retailer bigint,
    bonus_discount_stake numeric(10,2),
    settlement_notification_type character varying(30),
    retailer_id bigint,
    brand_id smallint NOT NULL,
    test boolean DEFAULT false NOT NULL,
    mts boolean,
    refund_tax_on_returns_amt numeric(10,4),
    extra_info json,
    estimated_reward_points numeric(10,2),
    pools_betting_coupon_id bigint,
    settlement_version smallint DEFAULT 0 NOT NULL,
    arena_bet_type_id smallint DEFAULT 0 NOT NULL,
    mts_rejection_message character varying(200),
    geocomply_details character varying(200),
    teaser_price numeric(7,3),
    price_chan_id bigint,
    limits_chan_id bigint,
    tax_federal numeric(7,3),
    tax_state numeric(7,3),
    w2g_id bigint,
    geolocation_details character varying(200),
    mts_settlement boolean DEFAULT false
);

alter table bets1 replica identity full;
```

```sql
COPY bets1 from STDIN CSV;
213231430000103,0,1,2021-08-14 03:21:26.57717+01,site=NORMAL;ip-address=172.18.8.99;application=web-sportsbook;,ACCEPTED,MULTIPLE,2021-08-13 10:43:56.844+01,100.00,33,,,,0.00,0.00,INTERNET,USD,,5,33,0,PRE_MATCH_MULTIPLE,,0,0,,,0,2,1,3.5340,,,0,,,0,,,3,,1,,100.00,0,,,,,f,,US ,,,,1213231430000103,,,1,1,,2,,,0,,2,,f,,,,,0.00,,,,,,,,,t,,0,90.0000,3,,,,,f,,,0,0,,,f,f,499.96,,,,,,,,,,,2,t,f,,"{""bonuses"":[{""type"":""ColumnLengthBetBonus"",""amount"":0,""awarded"":false,""updatedOn"":""2021-08-13T09:43:57.078+0000""}]}",,,0,0,,,,1,5,,,,,f
\.

update bets1 set id=id;
```

I tried to debug it, and I believe the error come from this function's return

```py
    def process_update(self, message: ReplicationMessage, transaction: Transaction) -> ChangeEvent:
        decoded_msg: decoders.Update = decoders.Update(message.payload)
        relation_id: int = decoded_msg.relation_id
        if decoded_msg.old_tuple:
            before = map_tuple_to_dict(tuple_data=decoded_msg.old_tuple, relation=self.table_schemas[relation_id])
        after = map_tuple_to_dict(tuple_data=decoded_msg.new_tuple, relation=self.table_schemas[relation_id])
        logging.debug (f"before={self.table_models[relation_id](**before)}")  # HERE
        logging.debug (f"after={self.table_models[relation_id](**after)}")
        return ChangeEvent(
            op=decoded_msg.byte1,
            message_id=message.message_id,
            lsn=message.data_start,
            transaction=transaction,
            table_schema=self.table_schemas[relation_id],
            before=self.table_models[relation_id](**before) if decoded_msg.old_tuple else None,
            after=self.table_models[relation_id](**after),
        )
```

more specifically this part

```py
self.table_models[relation_id](**before)
```

**Comment by dgea005** [comment]
Hi @brunomgalmeida,

thank you for trying out the library and also for raising this issue 

The type casting from strings to python types is still pretty basic. This [function](https://github.com/dgea005/pypgoutput/blob/master/src/pypgoutput/reader.py#L87-L96) is used with dynamic models from [Pydantic](https://pydantic-docs.helpmanual.io/usage/models/#dynamic-model-creation) and I need to add handling of more of the Postgres types.

I'll look into this as soon as I have time (reproduce it, improve the error messages and fix it). I'm also considering an option to just return the before/after data as a string if it's useful.

What version of Postgres are you using?

**Comment by brunomgalmeida** [comment]
Hi @dgea005 

I have tested the above with postgres 10 and postgres 12, both yield a similar error, on update didnt test the insert.

> and I need to add handling of more of the Postgres types.

I guess I am using some data types that are not being handleded then

> I'm also considering an option to just return the before/after data as a string if it's useful.

I started doing this, but got sidetracked with work.... the killer of joy

Thanks for checking this

**Comment by dgea005** [comment]
Hey @brunomgalmeida,

I tried out your example on a local container and it worked perfectly to help debug it.

It seems I specified the pydantic type for `JSON` wrong. I put `dict` but there is a real `JSON` https://pydantic-docs.helpmanual.io/usage/types/

I merged a fix here in case you are interested to try using it again  https://github.com/dgea005/pypgoutput/pull/14

I only had to time to reply to you until now also because of work :smile: 

Thanks again for writing the issue!
